### ISSUE TITLE
Operator `satisfies` breaks contextual ThisType<T>'s marker

### ISSUE THREAD

**Issue by x6eull** [microsoft/TypeScript#57431]
### 🔎 Search Terms

satisfies ThisType

### 🕗 Version & Regression Information

I'v tested it in typescript v4.9.5 (`satisfies` was introduced in 4.9), v5.3.3 and v5.4.0-beta, and the problem seems to exist in these 3 versions.

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.3#code/MYewdgzgLgBCBGArGBeGBvAUDHMCGAXAIwAMANNrgGYExUCuYwUAluDABQCUGMokIADYBTAHSCQAcw5QAFiwii8PAL4wIeVhCothETjxQA+GADcQLACYVcdAEy0GTVu269+EIWInS5Cparqmgo6evhgAJ421ADMjozMbGAG7uCeIuJSMvKKyjAqmGp4+gAqOSURAA7CADzo+LRg9AC28MIATvlGANxAA

### 💻 Code

```ts
const obj = {
    a: 10,
    f: function () { console.log(this.a) } satisfies () => void,
    f2: function () { console.log(this.a) } satisfies any,
    f3: function () { console.log(this.a) }
} as ThisType<{ a: number }>;
```

### 🙁 Actual behavior

The complier reports that `'this' implicitly has type 'any' because it does not have a type annotation` in `obj.f` and `obj.f2` but not `obj.f3`.

### 🙂 Expected behavior

Not error should be reported since ThisType\<T\> has been applied and `satisfies` shouldn't break that marker.

**Comment by fatcerberus** [comment]
This seems expected because `satisfies` gives the function expression a new contextual type that doesn't include the `ThisType`.

**Comment by Andarist** [comment]
Just a somewhat fun example when we kinda can see how the object is typechecked with both contextual types... but separately:
```ts
const test = {
  a: [], // Object literal may only specify known properties, and 'a' does not exist in type '{ b: "foo"; }'.(2353)
  b: "foo", // Object literal may only specify known properties, and 'b' does not exist in type '{ a: []; }'.(2353)
} satisfies {
  a: [];
} satisfies {
  b: "foo";
};
```

**Comment by fatcerberus** [comment]
Wait, does that run two separate excess property checks, or do the contextual types actually merge there?

**Comment by Andarist** [comment]
There is only one contextual type here if I'm not mistaken. However, each `satisfies` operator typechecks the expression type against its own target type. And the result is kinda weird... :p 

The very same thing happens in other situations, like this one:
```ts
const test2: {
  a: [];
} = {
  a: [], // Object literal may only specify known properties, and 'a' does not exist in type '{ b: "foo"; }'.(2353)
  b: "foo", // Object literal may only specify known properties, and 'b' does not exist in type '{ a: []; }'.(2353)
} satisfies {
  b: "foo";
};
```

It's such an edge case that it's probably not worth touching this at all 😉 In fact, I can see the argument that this is very much desired. The object type here is still fresh so its still subject to the excess property check. It's just weird that you kinda have multiple contextual-like targets but not all of them actually contribute to the contextual typing of this object.

**Comment by x6eull** [comment]
I haven't found any document about `satisfies` on https://www.typescriptlang.org/, and I was regarding it as an operator to "assert the type without changing the expression's exact type".
Maybe an official introduction/definition would be a favor😊

**Comment by snarbies** [comment]
I think [4.9 release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator) is the only documentation available. The behavior doesn't really match the expectation the documentation sets up, i.e. that a satisfies clause should perform a type _check_ but otherwise is a no-op.

**Comment by fatcerberus** [comment]
@x6eull Yeah, the documentation isn't 100% clear about this - `satisfies` *normally* doesn't change the type of an expression, *except* through contextual typing (a mechanism that guides type inference according to the type that's being assigned to).  So for example:
1.  `[ 1, 2, 3 ]` is normally `number[]`, but `[ 1, 2, 3 ] satisfies [ number, number, number ]` infers a tuple type instead.
2. `x => x` is normally an implicit `any`, but `(x => x) satisfies (x: string) => string` succeeds and types `x` as `string`.

Unfortunately, `ThisType` also works through contextual typing, so your use of `satisfies` inside the contextually-typed object literal seems to prevent it from working properly by overriding the outer contextual type.

[Playground for illustration](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZiEBGGBeGBtGSA0MBMeAzDALoDcAUAPTUz0wB6A-JZaJLAiPmptnkIwSpGBACGUAJYQ4UgKYR+YAK4BbAEbyATnlWade9Vu1kqtBk1btw0GBvHaU6ABQAPNAD4YbgJTm6BhY2DjsHbV5XD1RvPzFJGTlFGHcALjEobSkwAHNfLwys3IDLYKA)

@snarbies 

> So, just trying to understand, does "new contextual type" refer to the typing of the property value sub-expression that, in this case, apparently does not have the this type that the overall object has, hence the error?

Yes, that's correct.  `function () { ... } satisfies () => void` means that the function expression becomes contextually typed by `() => void` (which has no `this` type) instead of `ThisType<{ a: number }>`.

**Comment by x6eull** [comment]
Thanks everyone involved for the kind discussion and explanation😊
I would close this since this's expected.
